### DISCUSSION TITLE
[bug?] fields in update_all did not get mongoized

### DISCUSSION THREAD

**Discussion by midnight-wonderer** [mongodb/mongoid#5812] [General]
I am using Mongoid 8.1.5.

By declaring the following document:
~~~ruby
class BuggyModel
  include ::Mongoid::Document
  store_in collection: 'buggy_models'

  class MapperType
    MAPPING = {
      'A' => 1,
      'B' => 2,
    }.freeze
    INVERSE_MAPPING = MAPPING.invert.freeze

    class << self
      def mongoize(string)
        return string unless string.is_a?(::String)
        MAPPING[string]
      end

      def demongoize(integer)
        return integer unless integer.is_a?(::Integer)
        INVERSE_MAPPING[integer]
      end

      def evolve(hex_string)
        mongoize(hex_string)
      end
    end
  end

  field :_t, as: :testing, type: MapperType
end
~~~

I expect
~~~ruby
::BuggyModel.create({ _id: 1 })
::BuggyModel.update_all({ '$set': { testing: 'B' } })
~~~
to produce
~~~json
{ "_id": 1, "_t": 2 }
~~~
in the database.

Instead, it produces:
~~~json
{ "_id": 1, "_t": "B" }
~~~

Not sure if this is intentional, but the "B" value doesn't get mongoized.

Also, unrelated to the bug, I audited some code there.
[This line](https://github.com/mongodb/mongoid/blob/a35d481ef41ef738be1821d82bebbb52915ef576/lib/mongoid/atomic_update_preparer.rb#L28C52-L28C64) does not look right.
~~~ruby
(atomic_updates['$set'] ||= {})[key] = mongoize_for(key, klass, key, value)
~~~
should be
~~~ruby
(atomic_updates['$set'] ||= {})[key] = mongoize_for('$set', klass, key, value)
~~~
it does not cause any bugs now, though.
But it might in the future.

If you happen to touch some code in that file, you might consider using The Boy Scout Rule to make an edit there also.

**Comment by dem** [comment]
@midnight-wonderer we hit into this problem as well during our journey of mongoid upgrade :-)

Custom type classes are not mongoized in update_all with $set operator.

We have a solution:
For master - https://github.com/mongodb/mongoid/pull/5814
For 8.1 branch - https://github.com/mongodb/mongoid/pull/5815

Currently we are using monkey patch on 8.1.5:
```
module Mongoid
  module Extensions
    module Hash
      def value_for(operator, klass, key, value)
        case operator
        when "$rename" then value.to_s
        when "$addToSet", "$push" then value.mongoize
        else mongoize_for(operator, klass, key, value)
        end
      end
    end
  end
end
```

It solves problem for us (quite serious in terms of the app functionality). Hopefully it will help.

And thanks for pointing to the line, which does not looks right:
```
(atomic_updates['$set'] ||= {})[key] = mongoize_for(key, klass, key, value)
```

**Comment by alexbevi** [comment]
Thanks for raising this discussion @midnight-wonderer, and for the PR @dem. Our team will have a look and circle back.

**Comment by jamis** [comment]
Thank you very much. I've approved both PRs and am currently working through some infrastructure issues getting the 8.1-stable PR merged. Might have to wait until Monday for it. 👍
